File: bmsf-configuration/tools/healthz/healthz.go
/*
Tencent is pleased to support the open source community by making Blueking Container Service available.
Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except
in compliance with the License. You may obtain a copy of the License at
http://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied. See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"

	"github.com/olekukonko/tablewriter"
	"github.com/spf13/cobra"

	pbapiserver "bk-bscp/internal/protocol/apiserver"
	pbcommon "bk-bscp/internal/protocol/common"
	"bk-bscp/pkg/json"
)

var (
	// host is mysql host.
	host string

	// port is mysql port.
	port int

	// timeout is database read/write timeout.
	timeout time.Duration
)

func init() {
	flag.StringVar(&host, "host", "127.0.0.1", "APIServer host.")
	flag.IntVar(&port, "port", 8080, "APIServer port.")
	flag.DurationVar(&timeout, "timeout", 3*time.Second, "Health check request timeout.")
}

func healthz() (*pbapiserver.HealthzResponse, error) {
	client := &http.Client{Timeout: timeout}

	resp, err := client.Get(fmt.Sprintf("http://%s:%d/healthz", host, port))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	response := &pbapiserver.HealthzResponse{}
	if err := json.UnmarshalPB(string(body), response); err != nil {
		return nil, err
	}

	if !response.Result {
		return nil, fmt.Errorf("request to healthz api failed, %s", body)
	}
	if response.Code != pbcommon.ErrCode_E_OK {
		return nil, fmt.Errorf("healthz failed, errcode: %d, errmsg: %s", response.Code, response.Message)
	}

	return response, nil
}

// genSimpleHealthzCmd generates simple healthz command.
func genSimpleHealthzCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "SimpleHealthz",
		Short: "Simple healthz check command.",
		Run: func(cmd *cobra.Command, args []string) {
			response, err := healthz()
			if err != nil {
				fmt.Printf("make healthz request failed, %+v\n", err)
				return
			}
			fmt.Println(response.Data.IsHealthy)
		},
	}

	// flags.
	cmd.Flags().AddGoFlag(flag.CommandLine.Lookup("host"))
	cmd.Flags().AddGoFlag(flag.CommandLine.Lookup("port"))
	cmd.Flags().AddGoFlag(flag.CommandLine.Lookup("timeout"))

	return cmd
}

// genHealthzCmd generates healthz command.
func genHealthzCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "Healthz",
		Short: "Healthz check command.",
		Run: func(cmd *cobra.Command, args []string) {
			response, err := healthz()
			if err != nil {
				fmt.Printf("make healthz request failed, %+v\n", err)
				return
			}
			fmt.Printf("healthz result[%+v]:\n", response.Data.IsHealthy)

			table := tablewriter.NewWriter(os.Stdout)
			table.SetHeader([]string{"module", "version", "build_time", "git_hash", "is_healthy", "message"})

			for _, module := range response.Data.Modules {
				var line []string
				line = append(line, module.Module)
				line = append(line, module.Version)
				line = append(line, module.BuildTime)
				line = append(line, module.GitHash)
				line = append(line, fmt.Sprintf("%+v", module.IsHealthy))
				line = append(line, module.Message)
				table.Append(line)
			}
			table.Render()
		},
	}

	// flags.
	cmd.Flags().AddGoFlag(flag.CommandLine.Lookup("host"))
	cmd.Flags().AddGoFlag(flag.CommandLine.Lookup("port"))
	cmd.Flags().AddGoFlag(flag.CommandLine.Lookup("timeout"))

	return cmd
}

// genSubCmds returns sub commands.
func genSubCmds() []*cobra.Command {
	cmds := []*cobra.Command{}
	cmds = append(cmds, genSimpleHealthzCmd())
	cmds = append(cmds, genHealthzCmd())
	return cmds
}

// bscp healthz tool.
func main() {
	// root command.
	rootCmd := &cobra.Command{Use: "bk-bscp-healthz-tool"}

	// sub commands.
	subCmds := genSubCmds()

	// add sub commands.
	rootCmd.AddCommand(subCmds...)

	// run root command.
	if err := rootCmd.Execute(); err != nil {
		os.Exit(1)
	}
}

File: bcs-ui/backend/utils/healthz.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
import logging

from django.http import JsonResponse

logger = logging.getLogger(__name__)

PREFIX = 'handler_'


class HealthzError(Exception):
    pass


def healthz_view(request):
    data = healthz_manager()
    result = {'data': data, 'result': True, 'message': ''}
    return JsonResponse(result)


def healthz_manager():
    data = {}
    handlers = filter(lambda x: x[0].startswith(PREFIX), globals().items())
    for handler in handlers:
        name = handler[0][len(PREFIX) :]
        func = handler[1]
        # 默认返回0
        data[name] = 0
        try:
            func()
        except HealthzError as error:
            data[name] = '%s' % error
        except Exception as error:
            data[name] = '%s' % error
            logger.exception('healthz handler %s error', name)
    return data


def handler_redis():
    from backend.utils.cache import rd_client

    if rd_client.set("__healthz__", 1) is False:
        raise HealthzError("redis set command failed")
    if rd_client.get("__healthz__") != b'1':
        raise HealthzError("redis get command failed")


def handler_gcs_mysql():
    from backend.container_service.projects.models import ProjectUser

    ProjectUser.objects.first()


def main():
    print(healthz_manager())


def test_sentry(request):
    a = int('')
    return JsonResponse({'test': a})


if __name__ == '__main__':
    main()
